olm/open-liberty-crd.yaml

```diff
@@ -16,4 +16,463 @@ spec:
   versions:
   - name: v1alpha1
     served: true
-    storage: true
\ No newline at end of file
+    storage: true
+  validation:
+    openAPIV3Schema:
+      properties:
+        spec:
+          properties:
+            arch:
+              properties:  
+                amd64:
+                  enum:
+                  - "0 - Do not use"
+                  - "1 - Least preferred"
+                  - "2 - No preference"
+                  - "3 - Most preferred"
+                  type: string        
+                ppc64le:
+                  enum:
+                  - "0 - Do not use"
+                  - "1 - Least preferred"
+                  - "2 - No preference"
+                  - "3 - Most preferred"
+                  type: string        
+                s390x:
+                  enum:
+                  - "0 - Do not use"
+                  - "1 - Least preferred"
+                  - "2 - No preference"
+                  - "3 - Most preferred"
+                  type: string                      
+              required:           
```

Comment:
Please keep in mind that I know next to nothing about this validation code :-)
when this says `required` does this mean that one must code the architecture preference in their CR instance?  Or will it pull the default values from `values.yaml` if they are not specified?

Reply:
When I was playing with this it seems that "required" is sensitive to its scope.   So, from the point that you already encountered the containing spec.arch then "amd64, ppc64le, and s390x" are required.

Reply:
Mike, I have the same question as Tim,  have you done any testing with this validation setup?  If setting required for all these fields means that they have to be supplied in the CR then I think we might want to remove the required attribute.  My view is that if you do supply a value in the CR spec you are `overriding` the defaults set in the values.yaml

Reply:
I tested 9x% of the added logic (I don't want to commit to 100% -- it took a few days to do this).  I tested by adding the validation logic to the CRD then changing various settings in the full_cr.yaml.  I set the values to the wrong type as well as commented them out.   From my testing I concluded the behavior of  "required" to be what I responded above.   It appears that the required is scoped.  In the case of the "spec.arch" property the "required" for "amd64, ppc64le, and s390x" are scoped under the properties of "arch"; meaning that they are required if you specify "spec.arch".   They are not required if you specify "spec".

I just retested this for the following scenarios.   If I update spec.arch to the following:
  arch:
    amd64: "2 - No preference"
    #ppc64le: "2 - No preference"
    #s390x: "2 - No preference"
I'll get a failure complaining of missing properties for arch:
spec.arch.ppc64le in body is required
spec.arch.s390x in body is required

Update to:
  #arch:
    #amd64: "2 - No preference"
    #ppc64le: "2 - No preference"
    #s390x: "2 - No preference"
kubectl apply -f deploy/full_cr.yaml  is fine:
openliberty.openliberty.io/demo3-openliberty created

Update to:
arch:
    #amd64: "2 - No preference"
    #ppc64le: "2 - No preference"
    #s390x: "2 - No preference"
After apply, you get:
spec.arch in body must be of type object: "null"

Reply:
Joe, the way I understand this to work is that the validation logic is used to accept the CR instance to the system (kubernetes validates that the CR respects the CRD validation logic).  Then separate from that, the operator notices that there is a new CR instance, and passes that off to Helm to create the dependent kubernetes objects (deployment, service account, etc) representing the new CR instance.  The values in the CR instance would over-ride the values in values.yaml, but the validation logic in the CRD is not aware of the values.yaml at all.

Reply:
Tim, I am with you on that,  I guess my question is can i submit a CR where the spec section is empty if i have required sections?
```
apiVersion: openliberty.io/v1alpha1
kind: OpenLiberty
metadata:
  name: demo3-openliberty
spec: 
```

Reply:
You can have this:
```yaml
apiVersion: openliberty.io/v1alpha1
kind: OpenLiberty
metadata:
  name: demo3-openliberty
spec: 
```
Which deploys fine.  You cannot have this:
```yaml
apiVersion: openliberty.io/v1alpha1
kind: OpenLiberty
metadata:
  name: demo3-openliberty
spec:
  arch:
```
because the `arch:` exists but is not a valid format.  You get this error:
```
The OpenLiberty "demo3-openliberty" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"openliberty.io/v1alpha1", "kind":"OpenLiberty", "metadata":map[string]interface {}{"creationTimestamp":"2019-05-15T17:11:42Z", "generation":1, "name":"demo3-openliberty", "namespace":"default", "uid":"82a2559e-7734-11e9-95fc-005056aae27b"}, "spec":map[string]interface {}{"arch":interface {}(nil)}}: validation failure list:
spec.arch in body must be of type object: "null"
```
If you try this:
```yaml
apiVersion: openliberty.io/v1alpha1
kind: OpenLiberty
metadata:
  name: demo3-openliberty
spec:
  arch:
    amd64: "2 - No preference"
    ppc64le: "2 - No preference"
```
you get this:
```
The OpenLiberty "demo3-openliberty" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"openliberty.io/v1alpha1", "kind":"OpenLiberty", "metadata":map[string]interface {}{"creationTimestamp":"2019-05-15T17:14:48Z", "generation":1, "name":"demo3-openliberty", "namespace":"default", "uid":"f1aecd3b-7734-11e9-95fc-005056aae27b"}, "spec":map[string]interface {}{"arch":map[string]interface {}{"amd64":"2 - No preference", "ppc64le":"2 - No preference"}}}: validation failure list:
spec.arch.s390x in body is required
```
because you haven't specified all of the required elements.  And finaly, if you specify this:
```yaml
apiVersion: openliberty.io/v1alpha1
kind: OpenLiberty
metadata:
  name: demo3-openliberty
spec:
  arch:
    amd64: "2 - No preference"
    ppc64le: "2 - No preference"
    s390x: "5 - No preference"
```
You get this:
```
The OpenLiberty "demo3-openliberty" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"openliberty.io/v1alpha1", "kind":"OpenLiberty", "metadata":map[string]interface {}{"creationTimestamp":"2019-05-15T17:15:42Z", "generation":1, "name":"demo3-openliberty", "namespace":"default", "uid":"11dac6b9-7735-11e9-95fc-005056aae27b"}, "spec":map[string]interface {}{"arch":map[string]interface {}{"amd64":"2 - No preference", "ppc64le":"2 - No preference", "s390x":"5 - No preference"}}}: validation failure list:
spec.arch.s390x in body should be one of [0 - Do not use 1 - Least preferred 2 - No preference 3 - Most preferred]
```
because `5 - No preference` is not a valid choice.